content/post/postgres/rfc-pgxn-metadata-sketch.md

```diff
@@ -0,0 +1,959 @@
+        dependencies. For example, to specify alternate dependencies for Linux
+        arm64, it would be `"platforms": { "linux": "arm64" }`. For a specific
+        version of Postgres, it would be something like
+        `"postgres": { "version": ">= 16, < 17" }`. [This likely needs some
+        deep thinking through.]
+    *   **dependencies**: An object containing of the other key structures in
+        this object, defining the additional dependencies for environments
+        that match the `where` field.
+
+The `packages` and `external` keys point to objects in which the keys are
+dependency names and the values are version range expressions (empty string
+means any version). The `postgres.version` field is also a version range
+expression.
+
+The `contrib` key points to an array of Postgres-supplied apps and extensions.
+Versions are not necessary because they effectively have the same version as a
+supported Postgres version.
+
+Example:
+
+``` json
+{
+  "postgres": {
+    "version": ">= 12, < 17",
+    "with": [ "xml", "uuid", "perl" ]
+  },
+  "pipeline": "pgrx",
+  "platforms": {
+    "linux": [ "amd64", "arm64" ],
+    "darwin": [ "amd64", "arm64" ],
+    "windows": [ "amd64" ],
+    "freebsd": [ "amd64" ]
+  },
+  "dependencies": {
+    "configure": {
+      "requires": {
+        "external": { "cargo-pgrx": "" }
+      }
+    },
+    "build": {
+      "requires": {
+        "external": {
+          "curl": "",
+          "awk": "",
+          "sed": "",
+          "perl": "5.20"
+        }
+      },
+      "recommends": {
+        "external": {
+          "jq": "",
+          "perl": "5.38"
+        }
+      }
+    },
+    "test": {
+      "requires": {
+        "contrib": [ "pg_regress", "plpgsql", "plperl" ],
+        "packages": {
+          "github.com/theory/pgtap": "1.1.0"
+        }
+      }
+    },
+    "run": {
+      "requires": {
+        "contrib": [ "plperl" ],
+        "packages": {
+          "github.com/theory/hostname": ""
+        }
+      }
+    }
+  }
+}
+```
+
+**Notes:**
+
+The `external` field is the tricky one. How does one provide a canonical
+dependency for a third-party library or app that may have loads of different
+names on different platforms, different packaging systems (apt vs rpm), and
+even different versions of them, or split apart for development and runtime
+use? Some possibilities:
+
+*   Use [purl] and let downstream services handle the conversion
+*   Create a rule set that allows one to predict dependency name changes
+    across ecosystems, similar to [repology-rules]
+*   Just use [Repology]?
+
+Other notes:
+
+*   Decide on a version range standards to adopt. Start with the [PGXN version
+    range], [pep-0508], [npm ranges], and [Gem::Version].
+*   Is `pipeline` really necessary, given configure requirements? I think so,
+    because it tells the client the preferred build system to use, in case it
+    can't detect it for some reason.
```

Comment:
Sorry, I don't understand the question. But the point of `pipeline` is so that a client that downloads one of these sources packages knows what build pipeline to use to build it (including compilation, etc.).